source/simulator/assemblers/advection.cc

```diff
@@ -813,8 +817,12 @@ namespace aspect
              :
              cell->neighbor_face_no(face_no));
 
+          // TODO: need to consider subfaces on periodic neighbors of a coarsen cell
+          // as in such case, the face of current active cell is on the boundary and face->number_of_children()==1
+
           // loop over subfaces
-          for (unsigned int subface_no=0; subface_no<face->number_of_children(); ++subface_no)
+          typename DoFHandler<dim>::face_iterator neighbor_face=neighbor->face(neighbor2);
+          for (unsigned int subface_no=0; subface_no<neighbor_face->number_of_children(); ++subface_no)
```

Comment:
yes, this looks correct to me. maybe expand the comment to say
```
  // loop over subfaces. we know that the neighbor is finer, so we could loop over the subfaces of the current
  // face. but if we are at a periodic boundary, then the face of the current cell has no children, so instead use
  // the children of the periodic neighbor's corresponding face since we know that the letter does indeed have
  // children (because we know that the neighbor is refined).
```

By the way, you may also want to adjust the comment on the `else` in line 811 to reflect the new condition in the corresponding `if`.